src/lib/falcor/falcorUtils.js

```diff
@@ -123,11 +122,69 @@ export function expandCache(cache) {
     }, cache);
   }
 
+  function assignByPath(path, value) {
+    // If using dot notation obj.key.key.key
+    if (typeof path === "string") {
+      path = path.split('.');
+    }
+    // Parent also works for array length 1, aka initial keys
+    // Parent and Key variables are used for assigning new values later
+    const parent = followPath(path.slice(0, path.length-1));
+    const key = path[path.length-1];
+    // The following key exists as it was pushed on to stack as a valid key
+    parent[key] = value;
+  }
+
   function isObject(val) {
+    // We don't count arrays as objects here. This is to protect ourselves against an expanded atom
+    // This does still leave us vulnerable to an expanded object though, but in by far most cases
+    // it would be very bad form to put an object in an atom, so this is not supported at this time.
     if (val === null || (val instanceof Array)) return false;
     return typeof val === "object";
   }
 
+  function handleRef(pathArray, refPath) {
+    const S = new Set();
+    S.add(pathArray.join('.'));
+    let val = followPath(refPath);
+    let path = refPath.join('.');
+    if (val === undefined) {
+      throw new Error("Missing part of JSON graph in expandCache function at path: " + path);
+    }
+    while (isObject(val) && val.$type) {
+      switch (val.$type) {
+        case "atom":
+          assignByPath(path, val.value);
+          val = followPath(path);
+          break;
+        case "error":
+          assignByPath(path, new Error(val.value));
+          val = followPath(path);
+          break;
+        case "ref":
+          if (S.has(path)) {
+            let paths = "[";
+            S.forEach((pathFromSet) => {
+              paths += pathFromSet + ", ";
+            });
+            paths = paths.substring(0, paths.length-2) + ']';
+            throw new Error("Empty loop of circular refs at the following paths: " + paths)
```

Comment:
I don't print the specific path because I think it's very relevant knowing the whole loop. It might not necessarily be the exact place that the loop "closed" that'll be relevant, as the problem might be somewhere else.

And with the message, the reason I use "Empty Loop" is because we do support circular refs, the problem is when it's just refs that constantly refer to each other. 

What do you think I should end up doing?